File: RB101/small_problems/medium_1/3.rb
def rotation(array)
  array[1..-1] + [array[0]]
end

def rotate_rightmost_digits(number, n)
  result = number.to_s.chars
  result[-n..-1] = rotation(result[-n..-1])
  result.join.to_i
end

def max_rotation(num)
  num.digits.size.downto(2) do |i|
    num = rotate_rightmost_digits(num, i)
  end
  num
end

p max_rotation(735291) == 321579
p max_rotation(3) == 3
p max_rotation(35) == 53
p max_rotation(105) == 15 # the leading zero gets dropped
p max_rotation(8_703_529_146) == 7_321_609_845

File: JS210/small_problems/medium1/3.js
function rotateArray(arr) {
  if (!Array.isArray(arr)) return undefined;
  return arr.length > 0 ? arr.slice(1).concat(arr[0]) : [];
}

function rotateRightmostDigits(num, idx) {
  let arr = String(num).split('').map(Number);

  let firstHalf = arr.slice(0, arr.length - idx)
  let secondHalf = rotateArray(arr.slice(-idx));

  return Number([...firstHalf, ...secondHalf].join(''));
}

function maxRotation(number) {
  let digits = String(number).length;

  for (let i = digits; i > 1; i -= 1) {
    number = rotateRightmostDigits(number, i);
  }

  return number;
}

console.log(maxRotation(735291));          // 321579
console.log(maxRotation(3));               // 3
console.log(maxRotation(35));              // 53
console.log(maxRotation(105));             // 15 -- the leading zero gets dropped
console.log(maxRotation(8703529146));      // 7321609845
